File: packages/core-py/src/mcp_relay_core/relay/browser.py
"""Cross-platform browser opening with WSL detection."""

import logging
import subprocess
import webbrowser

logger = logging.getLogger(__name__)


def _is_wsl() -> bool:
    """Detect if running inside WSL."""
    try:
        with open("/proc/version", encoding="utf-8") as f:
            version = f.read().lower()
        return "microsoft" in version or "wsl" in version
    except OSError:
        return False


def _open_in_wsl(url: str) -> bool:
    """Open URL from inside WSL using wslview or cmd.exe."""
    # Try wslview first (from wslu package, commonly available)
    try:
        subprocess.run(
            ["wslview", url],
            check=True,
            capture_output=True,
            timeout=10,
        )
        return True
    except (FileNotFoundError, subprocess.SubprocessError):
        pass

    # Fallback to rundll32.exe url.dll,FileProtocolHandler
    try:
        subprocess.run(
            ["rundll32.exe", "url.dll,FileProtocolHandler", url],
            check=True,
            capture_output=True,
            timeout=10,
        )
        return True
    except (FileNotFoundError, subprocess.SubprocessError):
        pass

    return False


def try_open_browser(url: str) -> bool:
    """Try to open URL in default browser. Returns True if likely succeeded.

    Detection order:
    1. WSL: check /proc/version for Microsoft/WSL, use 'wslview' or 'cmd.exe /c start'
    2. Standard: webbrowser.open()

    Never raises. Returns False on failure.

    Args:
        url: The URL to open.

    Returns:
        True if the browser was likely opened, False otherwise.
    """
    try:
        # 1. WSL detection
        if _is_wsl():
            logger.debug("WSL detected, using WSL-specific browser opening")
            result = _open_in_wsl(url)
            if result:
                return True
            logger.debug("WSL browser opening failed, falling through to webbrowser")

        # 2. Standard webbrowser
        result = webbrowser.open(url)
        if result:
            logger.debug("Opened browser via webbrowser.open()")
        else:
            logger.debug("webbrowser.open() returned False")
        return result

    except Exception as err:
        logger.debug("Failed to open browser: %s", err)
        result = False

    if not result:
        import sys

        banner = f"""
\x1b[93m╔{"═" * 78}╗
║  \x1b[91mACTION REQUIRED: Browser auto-open failed.\x1b[93m {" " * 33}║
║  \x1b[97mPlease manually open this URL to continue setup:\x1b[93m {" " * 27}║
║  \x1b[36m{url:{74}s}\x1b[93m  ║
╚{"═" * 78}╝\x1b[0m
"""
        print(banner, file=sys.stderr)

    return result

File: packages/core-ts/src/relay/browser.ts
/**
 * Cross-platform browser opening with WSL detection.
 */

import { execFile } from 'node:child_process'
import { readFile } from 'node:fs/promises'
import { promisify } from 'node:util'

const execFileAsync = promisify(execFile)

async function isWsl(): Promise<boolean> {
  try {
    const version = await readFile('/proc/version', 'utf-8')
    const lower = version.toLowerCase()
    return lower.includes('microsoft') || lower.includes('wsl')
  } catch {
    return false
  }
}

async function openInWsl(url: string): Promise<boolean> {
  // Try wslview first (from wslu package, commonly available)
  try {
    await execFileAsync('wslview', [url])
    return true
  } catch {
    /* fall through */
  }

  // Fallback to rundll32.exe (secure alternative to cmd.exe /c start)
  try {
    await execFileAsync('rundll32.exe', ['url.dll,FileProtocolHandler', url])
    return true
  } catch {
    /* fall through */
  }

  return false
}

/**
 * Try to open URL in default browser. Returns true if likely succeeded.
 *
 * Detection order:
 * 1. win32: `rundll32.exe url.dll,FileProtocolHandler`
 * 2. darwin: `open` command
 * 3. linux: check WSL then `xdg-open`
 *
 * Never throws. Returns false on failure.
 */
export async function tryOpenBrowser(url: string): Promise<boolean> {
  try {
    // Validate URL schema as a defense-in-depth measure
    if (!/^https?:\/\//i.test(url)) {
      return false
    }

    const platform = process.platform

    if (platform === 'win32') {
      await execFileAsync('rundll32.exe', ['url.dll,FileProtocolHandler', url])
      return true
    }

    if (platform === 'darwin') {
      await execFileAsync('open', [url])
      return true
    }

    // linux
    if (await isWsl()) {
      const result = await openInWsl(url)
      if (result) return true
      // Fall through to xdg-open
    }

    await execFileAsync('xdg-open', [url])
    return true
  } catch {
    return false
  }
}
